#1. 기본 문법 익히기

In [33]:
from sklearn.neighbors import KNeighborsClassifier

# clf = KNeighborsClassifier(n_neighbors=5)
clf = KNeighborsClassifier(n_neighbors=3)

In [34]:
X_train = [
           [0.5,0.2,0.1],
           [0.9,0.7,0.3],
           [0.4,0.5,0.7]
]
y_train = [0,1,1]
clf.fit(X_train, y_train)

KNeighborsClassifier(algorithm='auto', leaf_size=30, metric='minkowski',
                     metric_params=None, n_jobs=None, n_neighbors=3, p=2,
                     weights='uniform')

In [35]:
X_test =[
         [0.2,0.1,0.4],
         [0.7,0.3,0.1],
         [0.5,0.5,0.5]
]
clf.predict(X_test)

array([1, 1, 1])

> 데이터의 개수보다 최근접 이웃의 개수가 적어야 함으로 n_neighbors를 조절해야 한다. 즉 n_neighbors는 feature의 개수보다 적게 작성해야 한다.

#2. 유방암 데이터 분석

##2.1 데이터 읽기

In [36]:
from sklearn.datasets import load_breast_cancer
breast_cancer = load_breast_cancer()
print(breast_cancer['DESCR'])

.. _breast_cancer_dataset:

Breast cancer wisconsin (diagnostic) dataset
--------------------------------------------

**Data Set Characteristics:**

    :Number of Instances: 569

    :Number of Attributes: 30 numeric, predictive attributes and the class

    :Attribute Information:
        - radius (mean of distances from center to points on the perimeter)
        - texture (standard deviation of gray-scale values)
        - perimeter
        - area
        - smoothness (local variation in radius lengths)
        - compactness (perimeter^2 / area - 1.0)
        - concavity (severity of concave portions of the contour)
        - concave points (number of concave portions of the contour)
        - symmetry 
        - fractal dimension ("coastline approximation" - 1)

        The mean, standard error, and "worst" or largest (mean of the three
        largest values) of these features were computed for each image,
        resulting in 30 features.  For instance, field 3 is Mean Radius, f

In [37]:
import pandas as pd

df_data = pd.DataFrame(breast_cancer.data)
df_labels = pd.DataFrame(breast_cancer.target)

##2.2 데이터 정제

In [38]:
df_data.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 569 entries, 0 to 568
Data columns (total 30 columns):
 #   Column  Non-Null Count  Dtype  
---  ------  --------------  -----  
 0   0       569 non-null    float64
 1   1       569 non-null    float64
 2   2       569 non-null    float64
 3   3       569 non-null    float64
 4   4       569 non-null    float64
 5   5       569 non-null    float64
 6   6       569 non-null    float64
 7   7       569 non-null    float64
 8   8       569 non-null    float64
 9   9       569 non-null    float64
 10  10      569 non-null    float64
 11  11      569 non-null    float64
 12  12      569 non-null    float64
 13  13      569 non-null    float64
 14  14      569 non-null    float64
 15  15      569 non-null    float64
 16  16      569 non-null    float64
 17  17      569 non-null    float64
 18  18      569 non-null    float64
 19  19      569 non-null    float64
 20  20      569 non-null    float64
 21  21      569 non-null    float64
 22  22

> 총 30개의 feature 정보를 갖고 있다.

In [39]:
df_data.head()

0      1       2       3   ...      26      27      28       29
0  17.99  10.38  122.80  1001.0  ...  0.7119  0.2654  0.4601  0.11890
1  20.57  17.77  132.90  1326.0  ...  0.2416  0.1860  0.2750  0.08902
2  19.69  21.25  130.00  1203.0  ...  0.4504  0.2430  0.3613  0.08758
3  11.42  20.38   77.58   386.1  ...  0.6869  0.2575  0.6638  0.17300
4  20.29  14.34  135.10  1297.0  ...  0.4000  0.1625  0.2364  0.07678

[5 rows x 30 columns]

> 현재 데이터가 어떤 값은 1보다 큰 값으로 어떤 값은 1보다 작은 값으로 표현되어 이를 정규화(일반화) 시킬 것이다.

In [40]:
df_data.describe()

0           1           2   ...          27          28          29
count  569.000000  569.000000  569.000000  ...  569.000000  569.000000  569.000000
mean    14.127292   19.289649   91.969033  ...    0.114606    0.290076    0.083946
std      3.524049    4.301036   24.298981  ...    0.065732    0.061867    0.018061
min      6.981000    9.710000   43.790000  ...    0.000000    0.156500    0.055040
25%     11.700000   16.170000   75.170000  ...    0.064930    0.250400    0.071460
50%     13.370000   18.840000   86.240000  ...    0.099930    0.282200    0.080040
75%     15.780000   21.800000  104.100000  ...    0.161400    0.317900    0.092080
max     28.110000   39.280000  188.500000  ...    0.291000    0.663800    0.207500

[8 rows x 30 columns]

##2.3 데이터 정규화

In [41]:
from sklearn.preprocessing import MinMaxScaler

scaler = MinMaxScaler()
# scaler.fit(df_data)
# data_scaled = scaler.transform(dt_data)
data_scaled = scaler.fit_transform(df_data)
df_data_scaled = pd.DataFrame(data_scaled)
# df_data_scaled.head()
df_data_scaled.describe()

0           1           2   ...          27          28          29
count  569.000000  569.000000  569.000000  ...  569.000000  569.000000  569.000000
mean     0.338222    0.323965    0.332935  ...    0.393836    0.263307    0.189596
std      0.166787    0.145453    0.167915  ...    0.225884    0.121954    0.118466
min      0.000000    0.000000    0.000000  ...    0.000000    0.000000    0.000000
25%      0.223342    0.218465    0.216847  ...    0.223127    0.185098    0.107700
50%      0.302381    0.308759    0.293345  ...    0.343402    0.247782    0.163977
75%      0.416442    0.408860    0.416765  ...    0.554639    0.318155    0.242949
max      1.000000    1.000000    1.000000  ...    1.000000    1.000000    1.000000

[8 rows x 30 columns]

> 정규화 작업을 위해 MinMaxScaler클래스를 사용하였다. 현재 feature의 개수를 맞추기 위해 fit함수를 이용하여 데이터를 변환하기 위해 transform 함수를 이용한다. 이 두 가지는 항상 같이 동작시키기 때문에 두 기능을 합친 fit_transform 함수를 이용하였다. 이렇게 만들어진 데이터를 dataframe으로 변환한 후 head 및 describe로 0~1사이의 값으로 변환되었는지 확인하였다.

##2.4 모델 만들기

###2.4.1 데이터 분류

In [42]:
from sklearn.model_selection import train_test_split
X_train, X_test, y_train, y_test = train_test_split(df_data_scaled, df_labels, test_size=0.2, random_state=62)

##2.4.2 모델 생성

In [43]:
from sklearn.neighbors import KNeighborsClassifier

clf = KNeighborsClassifier()
clf.fit(X_train, y_train)

/usr/local/lib/python3.7/dist-packages/ipykernel_launcher.py:4: DataConversionWarning: A column-vector y was passed when a 1d array was expected. Please change the shape of y to (n_samples, ), for example using ravel().
  after removing the cwd from sys.path.


KNeighborsClassifier(algorithm='auto', leaf_size=30, metric='minkowski',
                     metric_params=None, n_jobs=None, n_neighbors=5, p=2,
                     weights='uniform')

##2.5 정확도 평가하기

In [44]:
clf.score(X_test, y_test)

0.956140350877193

##2.6 튜닝하기

In [45]:
from sklearn.neighbors import KNeighborsClassifier

knn_clf = KNeighborsClassifier()

In [46]:
from sklearn.model_selection import GridSearchCV

def  get_best_hyperparam():
  params ={
      'n_neighbors':n_neighbors_list,
      'weights':['uniform', 'distance'],
      'metric':['euclidean', 'manhattan', 'minkowski']
  }

  grid_cv = GridSearchCV(knn_clf, param_grid=params, scoring='accuracy', cv=5, verbose=1)
  grid_cv.fit(X_train, y_train)

  print('최고 점수 : ', grid_cv.best_score_)
  print('최적의 파라미터 : ', grid_cv.best_params_)

In [47]:
n_neighbors_list = range(1, 100, 10)
get_best_hyperparam()

[Parallel(n_jobs=1)]: Using backend SequentialBackend with 1 concurrent workers.
/usr/local/lib/python3.7/dist-packages/sklearn/model_selection/_validation.py:515: DataConversionWarning: A column-vector y was passed when a 1d array was expected. Please change the shape of y to (n_samples, ), for example using ravel().
  estimator.fit(X_train, y_train, **fit_params)
/usr/local/lib/python3.7/dist-packages/sklearn/model_selection/_validation.py:515: DataConversionWarning: A column-vector y was passed when a 1d array was expected. Please change the shape of y to (n_samples, ), for example using ravel().
  estimator.fit(X_train, y_train, **fit_params)
/usr/local/lib/python3.7/dist-packages/sklearn/model_selection/_validation.py:515: DataConversionWarning: A column-vector y was passed when a 1d array was expected. Please change the shape of y to (n_samples, ), for example using ravel().
  estimator.fit(X_train, y_train, **fit_params)
/usr/local/lib/python3.7/dist-packages/sklearn/model_selec

Fitting 5 folds for each of 60 candidates, totalling 300 fits


/usr/local/lib/python3.7/dist-packages/sklearn/model_selection/_validation.py:515: DataConversionWarning: A column-vector y was passed when a 1d array was expected. Please change the shape of y to (n_samples, ), for example using ravel().
  estimator.fit(X_train, y_train, **fit_params)
/usr/local/lib/python3.7/dist-packages/sklearn/model_selection/_validation.py:515: DataConversionWarning: A column-vector y was passed when a 1d array was expected. Please change the shape of y to (n_samples, ), for example using ravel().
  estimator.fit(X_train, y_train, **fit_params)
/usr/local/lib/python3.7/dist-packages/sklearn/model_selection/_validation.py:515: DataConversionWarning: A column-vector y was passed when a 1d array was expected. Please change the shape of y to (n_samples, ), for example using ravel().
  estimator.fit(X_train, y_train, **fit_params)
/usr/local/lib/python3.7/dist-packages/sklearn/model_selection/_validation.py:515: DataConversionWarning: A column-vector y was passed when

최고 점수 :  0.9648351648351647
최적의 파라미터 :  {'metric': 'euclidean', 'n_neighbors': 11, 'weights': 'uniform'}


/usr/local/lib/python3.7/dist-packages/sklearn/model_selection/_validation.py:515: DataConversionWarning: A column-vector y was passed when a 1d array was expected. Please change the shape of y to (n_samples, ), for example using ravel().
  estimator.fit(X_train, y_train, **fit_params)
/usr/local/lib/python3.7/dist-packages/sklearn/model_selection/_validation.py:515: DataConversionWarning: A column-vector y was passed when a 1d array was expected. Please change the shape of y to (n_samples, ), for example using ravel().
  estimator.fit(X_train, y_train, **fit_params)
/usr/local/lib/python3.7/dist-packages/sklearn/model_selection/_validation.py:515: DataConversionWarning: A column-vector y was passed when a 1d array was expected. Please change the shape of y to (n_samples, ), for example using ravel().
  estimator.fit(X_train, y_train, **fit_params)
/usr/local/lib/python3.7/dist-packages/sklearn/model_selection/_validation.py:515: DataConversionWarning: A column-vector y was passed when

> 다음과 같이 경고가 나타나는 이유는 fit 적용 시 y_train이 1차원 배열로 적용되어야 하기 때문이다.
- /usr/local/lib/python3.7/dist-packages/sklearn/model_selection/_validation.py:515: DataConversionWarning: A column-vector y was passed when a 1d array was expected. Please change the shape of y to (n_samples, ), for example using ravel().
  estimator.fit(X_train, y_train, **fit_params)

In [48]:
print(type(y_train))
y_train.shape

<class 'pandas.core.frame.DataFrame'>


(455, 1)

> 확인 결과 타입은 dataframe이며 2차원 데이터를 가지고 있다.

In [49]:
from sklearn.model_selection import GridSearchCV

def  get_best_hyperparam():
  params ={
      'n_neighbors':n_neighbors_list,
      'weights':['uniform', 'distance'],
      'metric':['euclidean', 'manhattan', 'minkowski']
  }

  grid_cv = GridSearchCV(knn_clf, param_grid=params, scoring='accuracy', cv=5, verbose=1)
  # y_train의 value 정보를 reshape을 이용하여 1차원배열로 변경
  grid_cv.fit(X_train, y_train.values.reshape(-1))

  print('최고 점수 : ', grid_cv.best_score_)
  print('최적의 파라미터 : ', grid_cv.best_params_)

In [50]:
# 1, 11, 21, ~~~~
n_neighbors_list = range(1, 100, 10)
get_best_hyperparam()

[Parallel(n_jobs=1)]: Using backend SequentialBackend with 1 concurrent workers.


Fitting 5 folds for each of 60 candidates, totalling 300 fits
최고 점수 :  0.9648351648351647
최적의 파라미터 :  {'metric': 'euclidean', 'n_neighbors': 11, 'weights': 'uniform'}


[Parallel(n_jobs=1)]: Done 300 out of 300 | elapsed:    2.2s finished


> 실행 결과 n_neighbors가 11일 때 최고의 점수를 얻게 되었다. 조금 더 자세한 값을 얻기 위해 n_neighbors를 2 ~ 20 사이의 범위로 지정하여 결과를 확인해 보자.

In [51]:
# 2, 5, 8, ~~~~
n_neighbors_list = range(2, 20, 3)
get_best_hyperparam()

[Parallel(n_jobs=1)]: Using backend SequentialBackend with 1 concurrent workers.


Fitting 5 folds for each of 36 candidates, totalling 180 fits
최고 점수 :  0.9692307692307691
최적의 파라미터 :  {'metric': 'euclidean', 'n_neighbors': 8, 'weights': 'distance'}


[Parallel(n_jobs=1)]: Done 180 out of 180 | elapsed:    1.2s finished


> 확인 결과 n_neighbors는 8로 변경되었으며 weights가 uniform에서 distance로 변경된 것을 볼 수 있다. 마지막으로 6 ~ 10 사이의 수를 확인해 보자.

In [52]:
# 6, 7, 8, 9, 10
n_neighbors_list = range(6, 10, 1)
get_best_hyperparam()

[Parallel(n_jobs=1)]: Using backend SequentialBackend with 1 concurrent workers.


Fitting 5 folds for each of 24 candidates, totalling 120 fits
최고 점수 :  0.9692307692307691
최적의 파라미터 :  {'metric': 'euclidean', 'n_neighbors': 6, 'weights': 'distance'}


[Parallel(n_jobs=1)]: Done 120 out of 120 | elapsed:    0.8s finished


> 확인 결과 6이 최고의 점수를 갖는 파라미터로 나타 났다. 하지만 8과 6 둘 다 최고점은 같은 값으로 나타났다. 그러면 7도 같은 값을 갖게 되는지 확인해 보자

In [53]:
# 6, 7, 8, 9, 10
n_neighbors_list = [7]
get_best_hyperparam()

[Parallel(n_jobs=1)]: Using backend SequentialBackend with 1 concurrent workers.


Fitting 5 folds for each of 6 candidates, totalling 30 fits
최고 점수 :  0.9626373626373625
최적의 파라미터 :  {'metric': 'manhattan', 'n_neighbors': 7, 'weights': 'uniform'}


[Parallel(n_jobs=1)]: Done  30 out of  30 | elapsed:    0.2s finished


> 확인 결과 6, 8보다 작은 점수가 나타난 것을 볼 수 있다. 따라서 모델에 적용할 하이퍼 파라미터는 다음과 같이 적용할 수 있다.
- 'metric': 'manhattan'
- 'n_neighbors': 6 또는 8 
- 'weights': 'uniform'

> n_neighbors가 너무 작게 되면 이상치 처리에 문제게 될 수 있으나 너무 크게 되면 오랜 시간이 걸릴 수 있어 적당한 수치를 선택하여 사용하면 된다. 여기서는 오랜 시간이 걸리진 않음으로 8이 가장 좋은 파라미터가 될 수 있다.

## 2.7 최적화 적용

In [54]:
from sklearn.neighbors import KNeighborsClassifier
# 0.956140350877193
# clf = KNeighborsClassifier()

clf = KNeighborsClassifier(metric='manhattan', n_neighbors= 8, weights= 'uniform')
clf.fit(X_train, y_train)

clf.score(X_test, y_test)

/usr/local/lib/python3.7/dist-packages/ipykernel_launcher.py:6: DataConversionWarning: A column-vector y was passed when a 1d array was expected. Please change the shape of y to (n_samples, ), for example using ravel().
  


0.9649122807017544

#3. 모델 저장하기

##3.1 메모리에 저장하기

In [55]:
import pickle
# clf를 saved_model에 저장
saved_model = pickle.dumps(clf)

> saved_model에 clf를 저장한다. 메모리에 저장한다는 것은 변수에 저장한다는 개념과 같다.

In [57]:
# 메모리에 저장된 데이터를 모델로 읽어들인다.
clf_from_pickle = pickle.loads(saved_model)

In [ ]:
print(clf.predict(X_test))
print(clf_from_pickle.predict(X_test))

##3.2 파일로 저장하기

In [59]:
from google.colab import drive
drive.mount('/content/drive')

Mounted at /content/drive


In [61]:
from sklearn.externals import joblib

save_model_filepath='/content/drive/MyDrive/9. K-최근접 이웃/model/breast_cancer_model_v1.pkl'
joblib.dump(clf, save_model_filepath)

/usr/local/lib/python3.7/dist-packages/sklearn/externals/joblib/__init__.py:15: FutureWarning: sklearn.externals.joblib is deprecated in 0.21 and will be removed in 0.23. Please import this functionality directly from joblib, which can be installed with: pip install joblib. If this warning is raised when loading pickled models, you may need to re-serialize those models with scikit-learn 0.21+.
  warnings.warn(msg, category=FutureWarning)


['/content/drive/MyDrive/9. K-최근접 이웃/model/breast_cancer_model_v1.pkl']

> save_model_filepath 위치에 모델 저장

In [63]:
filepath='/content/drive/MyDrive/9. K-최근접 이웃/model/X_test.csv'
X_test.to_csv(filepath)

> 모델을 읽어 들일 경우 테스트를 위한 데이터 저장

In [64]:
X_test.head(2)

0         1         2   ...        27        28        29
465  0.296228  0.352384  0.297699  ...  0.466323  0.252316  0.458219
274  0.518198  0.499493  0.493470  ...  0.390378  0.185098  0.160304

[2 rows x 30 columns]